spring-kafka/src/main/java/org/springframework/kafka/listener/FailedRecordTracker.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import java.util.function.BiConsumer;
+
+import org.apache.commons.logging.Log;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import org.springframework.lang.Nullable;
+
+/**
+ * Track record processing failure counts.
+ *
+ * @author Gary Russell
+ * @since 2.2
+ *
+ */
+class FailedRecordTracker {
+
+	private final ThreadLocal<FailedRecord> failures = new ThreadLocal<>(); // intentionally not static
```

Comment:
Gary, remind me, please, when we need it to be `static` and when not.

Thanks

Reply:
Per thread Vs. Per thread/instance: https://stackoverflow.com/questions/2784009/why-should-java-threadlocal-variables-be-static